frame/staking-rewards/src/lib.rs

```diff
@@ -494,11 +515,80 @@ pub mod pallet {
 
 		#[transactional]
 		fn unstake(
-			_who: &Self::AccountId,
-			_position: &Self::PositionId,
+			who: &Self::AccountId,
+			position_id: &Self::PositionId,
 			_remove_amount: Self::Balance,
 		) -> DispatchResult {
-			Err("Not implemented".into())
+			let keep_alive = false;
+			let stake = Stakes::<T>::try_get(position_id).map_err(|_| Error::<T>::StakeNotFound)?;
+			ensure!(who == &stake.owner, Error::<T>::OnlyStakeOwnerCanUnstake);
+			ensure!(!stake.reductions.is_empty(), Error::<T>::ReductionConfigProblem);
+			let early_unlock = stake.lock.started_at.safe_add(&stake.lock.duration)? >=
+				T::UnixTime::now().as_secs();
+			let pool_id = stake.reward_pool_id;
+			let mut rewards_pool =
+				RewardPools::<T>::try_get(pool_id).map_err(|_| Error::<T>::RewardsPoolNotFound)?;
+
+			let mut inner_rewards = rewards_pool.rewards.into_inner();
+			for (asset_id, reward) in inner_rewards.iter_mut() {
+				let inflation = match stake.reductions.get(asset_id) {
+					Some(inflation) => *inflation,
+					None => Zero::zero(),
+				};
+				let claim = if rewards_pool.total_shares == Zero::zero() {
+					Zero::zero()
+				} else {
+					reward
+						.total_rewards
+						.safe_mul(&stake.share)?
+						.safe_div(&rewards_pool.total_shares)?
+						.safe_sub(&inflation)?
+				};
+				let claim = if early_unlock {
+					(Perbill::one() - stake.lock.unlock_penalty).mul_ceil(claim)
+				} else {
+					claim
+				};
+				let claim = sp_std::cmp::min(
+					claim,
+					reward.total_rewards.safe_sub(&reward.claimed_rewards)?,
+				);
+				reward.claimed_rewards = reward.claimed_rewards.safe_add(&claim)?;
```

Comment:
no sure, but i doubt that rewards should be penalized, and if should, then claimed_rewards should be removed in fully for acounting, so transfer is part

Reply:
so when somebody claims and pealized, may be rewards should gone from pool fully, and only penalized retained in treasury.

Reply:
We don't penalize rewards here because they store in [total_rewards](https://github.com/ComposableFi/composable/blob/andor0/add-unstake-into-staking-rewards/frame/composable-traits/src/staking/mod.rs#L28).
While [claimed_rewards](https://github.com/ComposableFi/composable/blob/andor0/add-unstake-into-staking-rewards/frame/composable-traits/src/staking/mod.rs#L31) stores claimed rewards by stakers by unstaking.